packages/event-bus-redis/src/services/event-bus-redis.ts

```diff
@@ -0,0 +1,223 @@
+  queuePrefix?: string
+  workerPrefix?: string
+  redisUrl?: string
+}
+
+type BullJob<T> = {
+  data: {
+    eventName: string
+    data: T
+    completedSubscriberIds: string[] | undefined
+  }
+} & Job
+
+type EmitOptions = {
+  delay?: number
+  attempts: number
+  backoff?: {
+    type: "fixed" | "exponential"
+    delay: number
+  }
+}
+
+/**
+ * Can keep track of multiple subscribers to different events and run the
+ * subscribers when events happen. Events will run asynchronously.
+ */
+export default class RedisEventBusService extends AbstractEventBusModuleService {
+  protected readonly config_: ConfigModule
+  protected readonly moduleOptions_: EventBusRedisModuleOptions
+  protected readonly moduleDeclaration_: ConfigurableModuleDeclaration
+  protected readonly logger_: Logger
+
+  protected queue_: Queue
+
+  constructor(
+    { logger, configModule }: InjectedDependencies,
+    moduleOptions: EventBusRedisModuleOptions = {},
+    moduleDeclaration: ConfigurableModuleDeclaration
+  ) {
+    // @ts-ignore
+    super(...arguments)
+
+    if (moduleDeclaration?.resources !== MODULE_RESOURCE_TYPE.SHARED) {
+      throw new MedusaError(
+        MedusaError.Types.INVALID_ARGUMENT,
+        "At the moment this module can only be used with shared resources"
+      )
+    }
+
+    this.moduleOptions_ = moduleOptions
+    this.moduleDeclaration_ = moduleDeclaration
+    this.config_ = configModule
+    this.logger_ = logger
+
+    this.connect()
+  }
+
+  connect(): void {
+    // Required config
+    // See: https://github.com/OptimalBits/bull/blob/develop/CHANGELOG.md#breaking-changes
+    const connection = new Redis(this.moduleOptions_?.redisUrl!, {
+      maxRetriesPerRequest: null,
+    })
+
+    this.queue_ = new Queue(this.moduleOptions_.queueName ?? `events-queue`, {
+      connection,
+      prefix: this.moduleOptions_.queuePrefix ?? `${this.constructor.name}`,
+    })
+
+    // Register our worker to handle emit calls
+    new Worker(this.moduleOptions_.queueName ?? "events-queue", this.worker_, {
+      connection,
+      prefix: this.moduleOptions_.workerPrefix ?? `${this.constructor.name}`,
+    })
+  }
+
+  /**
+   * Calls all subscribers when an event occurs.
+   * @param {string} eventName - the name of the event to be process.
+   * @param data - the data to send to the subscriber.
+   * @param options - options to add the job with
+   * @return the job from our queue
+   */
+  async emit<T>(
+    eventName: string,
+    data: T,
+    options: Record<string, unknown> & EmitOptions = { attempts: 1 }
+  ): Promise<void> {
+    const opts: { removeOnComplete: boolean } & EmitOptions = {
+      removeOnComplete: true,
+      attempts: 1,
+    }
+    if (typeof options.attempts === "number") {
+      opts.attempts = options.attempts
+      if (isDefined(options.backoff)) {
+        opts.backoff = options.backoff
+      }
+    }
+    if (typeof options.delay === "number") {
+      opts.delay = options.delay
+    }
```

Comment:
question: Since this is a common feature for all event buses, shouldn't this be an abstraction in the base class?